### DISCUSSION TITLE
BIRT Report Engine 4.13 externalized data source not found

### DISCUSSION THREAD

**Discussion by nilsmork1** [eclipse-birt/birt#1397] [Q&A] [answered]
I created an Oracle JDBC data source for a simple cross-tab report.  I have used the "externalize data source" in order to share the connection setup with other reports in the project.
1. I verified the data source works in the BIRT All-In-One designer
2. I verified I can re-use the data source in a new report in the All-In-One designer
3. Data source is externalized to its own file (concept: copy any changes to password or host by copy the one file from dev to prod)
4. When I run the report using the BIRT report runtime engine with UN-externalized data source (data source info in rptdesign file only) report works
5. When I run the report using the BIRT report runtime engine, it does not find the externalized data source connection file & asks me to verify the path

Report engine is in C:\BIRT\reportEngine
Workspace environment variable %%org.eclipse.datatools_workspacepath is set to D:\cslink\BIRT
Workspace D:\cslink\BIRT is location for rptdesign files
Workspace D:\cslink\BIRT is location for externalized data source file (though I have also tried putting it into other locations, e.g. C:\BIRT\reportEngine\lib) 

I don't find any information about how to deploy the externalized data source XML to the reportEngine runtime.  Report Engine is called using Perl conversion of the bat file provided with the reportEngine distribution.

**Comment by speckyspooky** [comment]
I think you talk about the connection profile which is used on your data source.
Add the creation moment of a connection profile the path will be specified which is to be used to find the cp data source file

And I created for text reason an own Oracle-profile added it on a data source.
If I open the data source go to connection profile and open "Browse" with absolute path then the file target will be shown where I can find the file, see Image:
![connection-profile-folder](https://github.com/eclipse-birt/birt/assets/41593722/ab81907a-0429-4491-99e1-de83393b07ad)

Perhaps you can check the used folder (and file) on your side and the application later will use the same folder & file because the reference is the part of the data source so it is not dependend on BIRT it is dependend on your report and your data sources.

**Reply by nilsmork1** [reply, marked as answer]
Trial and error have resolved the issue.  I discovered by modifying the perl script to chdir to the directory where both the rptdesign file and the shared externalized data source file were, the report engine was able to find the shared data source file.

The problem occurred because I was using "relative path" in selecting the shared data source, also, since I assumed that the deployment server would have a very different directory tree from the development workstations, I forced it to have RELATIVE PATH and NO PATH (i.e. same directory as the rptdesign file.

The perl script launched from a different folder, and even though it had an environment variable to construct a full path for the rptdesign file and so could provide that to the BIRT report engine class, the BIRT report engine got the path for the shared data source file from the rptdesign xml... which HAD NO PATH WHATSOEVER saved.  So the report engine looked for the shared data source file in the same directory as the Perl script and found nothing.  Modifying the Perl script to chdir to the folder containing both the rptdesign file and the shared data source file before calling java to run the report engine solved the problem.

I hope future adventurers having similar problems can parse this gobbledygook I've written and benefit.
